gateway/ln-gateway/src/lib.rs

```diff
@@ -523,66 +584,100 @@ impl Gateway {
             .get_balance()
-            .await
+            .await)
     }
 
     async fn handle_address_msg(&self, payload: DepositAddressPayload) -> Result<Address> {
-        self.select_actor(payload.federation_id)
+        let (_, address) = self
+            .select_client(payload.federation_id)
             .await?
-            .get_deposit_address()
-            .await
-    }
-
-    async fn handle_deposit_msg(&self, payload: DepositPayload) -> Result<TransactionId> {
-        let DepositPayload {
-            txout_proof,
-            transaction,
-            federation_id,
-        } = payload;
-
-        self.select_actor(federation_id)
-            .await?
-            .deposit(txout_proof, transaction)
-            .await
+            .get_deposit_address(now() + Duration::from_secs(86400 * 365))
+            .await?;
+        Ok(address)
     }
 
-    async fn handle_withdraw_msg(&self, payload: WithdrawPayload) -> Result<TransactionId> {
+    async fn handle_withdraw_msg(&self, payload: WithdrawPayload) -> Result<Txid> {
         let WithdrawPayload {
             amount,
             address,
             federation_id,
         } = payload;
 
-        self.select_actor(federation_id)
+        let client = self.select_client(federation_id).await?;
+        // TODO: This should probably be passed in as a parameter
+        let fees = client.get_withdraw_fee(address.clone(), amount).await?;
+
+        let operation_id = self
+            .select_client(federation_id)
             .await?
-            .withdraw(amount, address)
-            .await
+            .withdraw(address, amount, fees)
+            .await?;
+        let mut updates = client
+            .subscribe_withdraw_updates(operation_id)
+            .await?
+            .into_stream();
+
+        while let Some(update) = updates.next().await {
+            info!("Update: {update:?}");
+
+            match update {
+                WithdrawState::Succeeded(txid) => {
+                    return Ok(txid);
+                }
+                WithdrawState::Failed(e) => {
+                    return Err(GatewayError::Other(anyhow!(e)));
+                }
+                _ => {}
+            }
+        }
+
+        return Err(GatewayError::Other(anyhow!(
+            "Unexpected error occurred while withdrawing"
+        )));
     }
 
     async fn handle_backup_msg(
         &self,
-        BackupPayload { federation_id }: BackupPayload,
+        BackupPayload { federation_id: _ }: BackupPayload,
     ) -> Result<()> {
-        self.select_actor(federation_id).await?.backup().await
+        unimplemented!("Backup is not currently supported");
     }
 
     async fn handle_restore_msg(
         &self,
-        RestorePayload { federation_id }: RestorePayload,
+        RestorePayload { federation_id: _ }: RestorePayload,
     ) -> Result<()> {
-        self.select_actor(federation_id).await?.restore().await
+        unimplemented!("Restore is not currently supported");
```

Comment:
Has backup/restore been implemented for client NG?

Reply:
Yes: https://github.com/fedimint/fedimint/blob/772c23c4dccdf2fa96fb508a1336a1725e9d98c0/fedimint-cli/src/ng.rs#L315C3-L325

Restore currently needs to be called when [initializing a client](https://github.com/fedimint/fedimint/blob/772c23c4dccdf2fa96fb508a1336a1725e9d98c0/fedimint-cli/src/lib.rs#L1086-L1106). We'd probably want to do roughly the same thing. Once it finishes, then just update the `clients` map to contain this new client as the value.

We'll have to be mindful about having 2 clients with same root secret for a federation.

Reply:
https://github.com/fedimint/fedimint/issues/2652